src/helper.cc

```diff
@@ -1198,17 +1198,11 @@ helperStatefulHandleRead(const Comm::ConnectionPointer &conn, char *, size_t len
     if (t) {
         /* end of reply found */
         srv->requests.pop_front(); // we already have it in 'r'
-        int called = 1;
-
-        if (r && cbdataReferenceValid(r->request.data)) {
-            r->reply.finalize();
-            r->reply.reservationId = srv->reservationId;
-            hlp->callBack(r);
-        } else {
-            debugs(84, DBG_IMPORTANT, "StatefulHandleRead: no callback data registered");
-            called = 0;
-        }
 
+        r->reply.finalize();
+        r->reply.reservationId = srv->reservationId;
+        const auto dataValid = cbdataReferenceValid(r->request.data);
```

Comment:
I am sorry, but after looking at these changes, I think we should restore the original code here. This refactoring does not make things better because finalize() does a lot of useless work when the caller is gone, and calling back with reply.reservationId when we are going to cancel that reservation looks like asking for trouble.

Let's keep this code structure unchanged, even though it is checking cbdataReferenceValid() twice, once here and once (wastefully for this caller) in callBack().